File: ios/ShiLv/App/AppModel.swift
import SwiftUI
import Photos
import UIKit

@MainActor
final class AppModel: ObservableObject {
    @Published var scanPhase: ScanPhase = .idle
    @Published var selectedTripID: UUID?
    @Published var presentedError: String?
    @Published private(set) var accessState: PhotoAccessState
    @Published private(set) var dataRevision = 0
    @Published var analysisProgress: (current: Int, total: Int)?
    @Published private(set) var thumbnailCacheSize: Int64 = 0
    @Published private(set) var isResettingIndex = false

    let photoLibrary: PhotoLibraryService
    let store: TripStore
    private let detector: TripDetector
    private let placeNames = PlaceNameService()
    private var activeScan: (id: UUID, task: Task<Void, Never>)?
    private var scanRequestedWhileBusy = false
    private var scanNeedsResume = false
    private var libraryChangeTask: Task<Void, Never>?

    init(photoLibrary: PhotoLibraryService = .shared, store: TripStore = TripStore(), detector: TripDetector = TripDetector()) {
        self.photoLibrary = photoLibrary
        self.store = store
        self.detector = detector
        self.accessState = photoLibrary.accessState
        photoLibrary.setChangeHandler { [weak self] in
            guard let self,
                  UIApplication.shared.applicationState == .active,
                  self.photoLibrary.accessState == .full || self.photoLibrary.accessState == .limited else { return }
            self.scheduleLibraryChangeScan()
        }
    }

    private func scheduleLibraryChangeScan() {
        libraryChangeTask?.cancel()
        libraryChangeTask = Task { @MainActor [weak self] in
            try? await Task.sleep(nanoseconds: 1_500_000_000)
            guard !Task.isCancelled, let self else { return }
            await self.scanLibrary()
        }
    }

    var trips: [DiscoveredTrip] { store.snapshot?.trips.filter { !$0.isHidden } ?? [] }
    func start() async {
        photoLibrary.refreshAccessState()
        accessState = photoLibrary.accessState
        if (accessState == .full || accessState == .limited), store.snapshot == nil {
            await scanLibrary()
        }
        await refreshStorageUsage()
    }

    func requestAndScan() async {
        guard scanPhase != .requestingPermission else { return }
        scanPhase = .requestingPermission
        let state = await photoLibrary.requestAccess()
        accessState = state
        guard state == .full || state == .limited else {
            scanPhase = .idle
            return
        }
        scanPhase = .idle
        await scanLibrary()
    }

    func scanLibrary() async {
        photoLibrary.refreshAccessState()
        accessState = photoLibrary.accessState
        guard accessState == .full || accessState == .limited else { scanPhase = .idle; return }
        if let active = activeScan {
            scanRequestedWhileBusy = true
            await active.task.value
            if activeScan?.id == active.id { activeScan = nil }
            return
        }
        let scanID = UUID()
        let task = Task { @MainActor [weak self] in
            guard let self else { return }
            await self.performScanLoop()
        }
        activeScan = (scanID, task)
        await task.value
        if activeScan?.id == scanID { activeScan = nil }
    }

    private func performScanLoop() async {
        repeat {
            scanRequestedWhileBusy = false
            await performSingleScan()
        } while scanRequestedWhileBusy && !Task.isCancelled && (accessState == .full || accessState == .limited)
    }

    private func performSingleScan() async {
        do {
            let records = await photoLibrary.fetchMetadata { [weak self] current, total in
                self?.scanPhase = .readingMetadata(current: current, total: total)
            }
            guard !Task.isCancelled else { scanPhase = .idle; return }
            photoLibrary.refreshAccessState()
            accessState = photoLibrary.accessState
            guard accessState == .full || accessState == .limited else { scanPhase = .idle; return }
            scanPhase = .detectingTrips
            let worker = Task.detached(priority: .userInitiated) { [detector] in
                detector.detect(from: records)
            }
            let snapshot = await withTaskCancellationHandler(operation: { await worker.value }, onCancel: { worker.cancel() })
            guard !Task.isCancelled else { scanPhase = .idle; return }
            scanPhase = .saving
            try store.replace(with: snapshot)
            dataRevision += 1
            scanPhase = .complete
        } catch {
            scanPhase = .failed("扫描失败，请稍后重试")
            presentedError = error.localizedDescription
        }
    }

    private var isScanning: Bool {
        switch scanPhase {
        case .requestingPermission, .readingMetadata, .detectingTrips, .saving: return true
        case .idle, .complete, .failed: return false
        }
    }

    func cancelActiveScan() {
        scanNeedsResume = isScanning
        scanRequestedWhileBusy = false
        libraryChangeTask?.cancel()
        activeScan?.task.cancel()
        if isScanning { scanPhase = .idle }
    }

    func resumeInterruptedScanIfNeeded() async {
        photoLibrary.refreshAccessState()
        accessState = photoLibrary.accessState
        guard accessState == .full || accessState == .limited else { scanNeedsResume = false; return }
        guard scanNeedsResume else { return }
        if let active = activeScan {
            await active.task.value
            if activeScan?.id == active.id { activeScan = nil }
        }
        scanNeedsResume = false
        await scanLibrary()
    }

    func update(_ trip: DiscoveredTrip) {
        var reconciled = trip
        reconciled.reconcileDerivedFields()
        do { try store.updateTrip(reconciled); dataRevision += 1 }
        catch { presentedError = "无法保存这次修改：\(error.localizedDescription)" }
    }

    func confirmAndAnalyze(_ trip: DiscoveredTrip) async -> DiscoveredTrip {
        analysisProgress = (0, max(1, trip.eventCount))
        let enriched = await MemoryAnalysisService(photoLibrary: photoLibrary).enrich(trip) { [weak self] current, total in
            self?.analysisProgress = (current, total)
        }
        guard !Task.isCancelled else { analysisProgress = nil; return trip }
        update(enriched)
        analysisProgress = nil
        return enriched
    }

    func resetIndex() async {
        guard !isResettingIndex else { return }
        isResettingIndex = true
        defer { isResettingIndex = false }
        cancelActiveScan()
        scanNeedsResume = false
        if let active = activeScan {
            await active.task.value
            if activeScan?.id == active.id { activeScan = nil }
        }
        do { try store.deleteLocalIndex(); dataRevision += 1; scanPhase = .idle }
        catch { presentedError = "无法删除本机索引：\(error.localizedDescription)" }
    }

    func refreshStorageUsage() async { thumbnailCacheSize = await photoLibrary.thumbnailCacheSize() }

    func clearThumbnailCache() async {
        do { try await photoLibrary.clearThumbnailCache(); await refreshStorageUsage() }
        catch { presentedError = "无法清理缩略图缓存：\(error.localizedDescription)" }
    }

    func resolveLocation(_ name: String) async -> GeoPoint? { await placeNames.locate(name) }

    func refreshAccessState() {
        photoLibrary.refreshAccessState()
        accessState = photoLibrary.accessState
    }
}

File: android/app/src/main/java/com/example/shilv/ui/AppModel.kt
package com.example.shilv.ui

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import android.graphics.Bitmap
import com.example.shilv.data.DiscoveredTrip
import com.example.shilv.data.GeoPoint
import com.example.shilv.data.PhotoAccessState
import com.example.shilv.data.PhotoRecord
import com.example.shilv.data.ScanPhase
import com.example.shilv.data.ScanSnapshot
import com.example.shilv.domain.TripDetector
import com.example.shilv.service.MemoryAnalysisService
import com.example.shilv.service.PhotoLibraryService
import com.example.shilv.service.PlaceNameService
import com.example.shilv.service.TripStore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.io.File

/** 全局状态：对应 iOS AppModel。 */
class AppModel(app: Application) : AndroidViewModel(app) {

    private val photoLibrary = PhotoLibraryService(app)
    private val store = TripStore(File(app.filesDir, "shilv-index"))
    private val detector = TripDetector()
    private val placeNames = PlaceNameService(app)

    private val _scanPhase = MutableStateFlow<ScanPhase>(ScanPhase.Idle)
    val scanPhase: StateFlow<ScanPhase> = _scanPhase.asStateFlow()

    private val _accessState = MutableStateFlow(photoLibrary.accessState)
    val accessState: StateFlow<PhotoAccessState> = _accessState.asStateFlow()

    private val _analysisProgress = MutableStateFlow<Pair<Int, Int>?>(null)
    val analysisProgress: StateFlow<Pair<Int, Int>?> = _analysisProgress.asStateFlow()

    private val _thumbnailCacheSize = MutableStateFlow(0L)
    val thumbnailCacheSize: StateFlow<Long> = _thumbnailCacheSize.asStateFlow()

    private val _dataRevision = MutableStateFlow(0)
    val dataRevision: StateFlow<Int> = _dataRevision.asStateFlow()

    private val _presentedError = MutableStateFlow<String?>(null)
    val presentedError: StateFlow<String?> = _presentedError.asStateFlow()

    val snapshot: ScanSnapshot? get() = store.snapshot
    val trips: List<DiscoveredTrip> get() = store.snapshot?.trips?.filter { !it.isHidden } ?: emptyList()
    val storedDataSize: Long get() = store.storedDataSize

    fun consumeError() { _presentedError.value = null }

    fun start() {
        _accessState.value = photoLibrary.accessState
        if (photoLibrary.accessState == PhotoAccessState.Full && store.snapshot == null) {
            scanLibrary()
        }
        refreshStorageUsage()
    }

    fun requestAndScan() {
        if (_scanPhase.value == ScanPhase.RequestingPermission) return
        _scanPhase.value = ScanPhase.RequestingPermission
        viewModelScope.launch {
            _accessState.value = photoLibrary.accessState
            if (photoLibrary.accessState != PhotoAccessState.Full) {
                _scanPhase.value = ScanPhase.Idle
                return@launch
            }
            _scanPhase.value = ScanPhase.Idle
            scanLibrary()
        }
    }

    fun scanLibrary() {
        if (photoLibrary.accessState != PhotoAccessState.Full) {
            _scanPhase.value = ScanPhase.Idle
            return
        }
        viewModelScope.launch {
            _scanPhase.value = ScanPhase.ReadingMetadata(0, 0)
            val records = photoLibrary.fetchMetadata { current, total ->
                _scanPhase.value = ScanPhase.ReadingMetadata(current, total)
            }
            if (photoLibrary.accessState != PhotoAccessState.Full) {
                _scanPhase.value = ScanPhase.Idle
                return@launch
            }
            _scanPhase.value = ScanPhase.DetectingTrips
            val snapshot = detector.detect(records)
            _scanPhase.value = ScanPhase.Saving
            store.replace(snapshot)
            _dataRevision.value += 1
            _scanPhase.value = ScanPhase.Complete
        }
    }

    fun confirmAndAnalyze(trip: DiscoveredTrip) {
        val eventCount = maxOf(1, trip.eventCount)
        _analysisProgress.value = 0 to eventCount
        viewModelScope.launch {
            val service = MemoryAnalysisService(photoLibrary, placeNames)
            val enriched = service.enrich(trip) { current, total ->
                _analysisProgress.value = current to total
            }
            update(enriched)
            _analysisProgress.value = null
        }
    }

    fun update(trip: DiscoveredTrip) {
        val reconciled = trip.also { it.reconcileDerivedFields() }
        store.updateTrip(reconciled)
        _dataRevision.value += 1
    }

    fun resetIndex() {
        viewModelScope.launch {
            store.deleteLocalIndex()
            _dataRevision.value += 1
            _scanPhase.value = ScanPhase.Idle
        }
    }

    fun refreshStorageUsage() {
        _thumbnailCacheSize.value = photoLibrary.thumbnailCacheSize()
    }

    fun clearThumbnailCache() {
        photoLibrary.clearThumbnailCache()
        refreshStorageUsage()
    }

    fun resolveLocation(name: String): GeoPoint? = null // 由调用方在协程中执行
    suspend fun resolveLocationAsync(name: String): GeoPoint? = placeNames.locate(name)

    suspend fun loadImage(id: String, targetSize: Int): Bitmap? =
        photoLibrary.requestImage(id, targetSize)
}

/** 供 UI 在协程中解析地点。 */
suspend fun AppModel.resolve(name: String): GeoPoint? = resolveLocationAsync(name)